### ISSUE TITLE
Editing a gene to specify an empty biotype crashes with HTTP 500 response

### ISSUE THREAD

**Issue by mgrbyte** [WormBase/names#72]
Re-produce by:
1) create new gene with a biotype and sequence name
2) edit the gene and set the biotype to empty.

**Comment by sibyl229** [comment]
@mgrbyte to reply to your comment quoted here:
> if you wanted, you could for e.g implement `ensureNoNullValues(payload)` as uniform transform for all POST/PUT

While it's technically possible, I have some usability related concerns. Changing a value to null is a deliberate user action, (even if it's done by mistake), so the user should be informed about the outcome of the action, especially when the outcome isn't what they would have expected. Without such a feedback, I imagine the user would expect the database record to reflect what they have set in the form, where certain value is set to null, which could be misleading.

I guess two approaches to address null issue: 1) not allow users to enter a null value if possible 2) display error message when null is entered. I think if the attribute is an ENUM, I can probably implement approach 1). I wonder how hard it is to provide validation error on the backend for other circumstances.

**Comment by mgrbyte** [comment]
Not sure why/what you mean regarding attribute being an ENUM and how that affects implementation?

Think 1) is the best route, since it seems counter-intuitive to allow a user to enter a invalid value.

The back-end end will need to validate that nil/null is not allowed, for any other usage (e.g scripts).

**Comment by sibyl229** [comment]
When it's an ENUM, the UI can ensure only valid options (of `<select>`) are available to users. 

If it's free form text, it will be done with some form of validation (either frontend or backend). 

I should be able to implement frontend validation, but if there will be backend validation anyway, then it's less urgent to implement frontend validation for 0.1 milestone, and I can spend the time on other items in 0.1 milestone. What do you think?

**Comment by mgrbyte** [comment]
I'm fine with deferring any niceties or features that aren't required back, but I think the following should work:
1) create un-cloned gene, edit it (e.g change cgc-name) and save

Even with back end validation in place, I don't think this will currently work without some handling of this nil/null issue?

**Comment by mgrbyte** [comment]
> While it's technically possible, I have some usability related concerns. Changing a value to null is a deliberate user action, (even if it's done by mistake), so the user should be informed about the outcome of the action, especially when the outcome isn't what they would have expected. Without such a feedback, I imagine the user would expect the database record to reflect what they have set in the form, where certain value is set to null, which could be misleading.

I've been thinking about your concerns here.
Initially, I didn't think that there could be a user-concern here, if : 
 - we changed the UI such that it was impossible for the user to enter a null value for a field (even by mistake)
 - and the server did valdiation, such that any other calls to the API would fail gracefully (400)

However, there is one case I can think of where this wouldn't work, which is the case I mention that _should_ work in my previous message (!)

tl;dr I think this needs more discussion/thought.
I think we should approach this from the UI angle (the details of the back-end, i.e datomic) shouldn't matter in terms of design of the UI (or REST API).

**Comment by sibyl229** [comment]
> Initially, I didn't think that there could be a user-concern here, if :
> - we changed the UI such that it was impossible for the user to enter a null value for a field (even by mistake)
> - and the server did valdiation, such that any other calls to the API would fail gracefully (400)

I agree.

Regarding the case you mentioned (and potentially other edge cases), can we clarify what are the conditions that undefined biotype is acceptable and what are the conditions that it's unacceptable?

**Comment by mgrbyte** [comment]
> Regarding the case you mentioned (and potentially other edge cases), can we clarify what are the conditions that undefined biotype is acceptable and what are the conditions that it's unacceptable?
Having thought this through and reviewed the code, I've found a issue that had confused me  with regard to this null issue.

There are no conditions where an `undefined/nil/empty` _biotype_ should be acceptable.

Turns out this _is_ a case where the `cgc-name` should be allowed to be unset from a gene though,
which is a similar issue.
In the later case (cgc-name), the server should accept null/empty values for cgc-name,
and eventually process them as retractions in datomic.

**Comment by sibyl229** [comment]
Thanks @mgrbyte ! Actually I meant to ask a different but related question: under what conditions is`undefined` a valid state in _the database_ (as apposed to the ajax payload)? The reason that this answer is important to me is that the UI models the state of the entity (like a mini-database on the frontend), the view of the form and the ajax payload are both derived from that. So knowing the possible states of an entity helps me figure out how to create the view. (I probably didn't explain this well...)

I seem to recall if biotype is never defined, it can stay undefined (in the database). Is that always true? Or do I need to consider the concept of cloned vs uncloned (where I have imagine the sequence name has a role)?

If the previous value is undefined, and a `undefined/null/empty` value is submitted in the payload. Does that attribute gets discarded?

**Comment by mgrbyte** [comment]
> `Thanks @mgrbyte ! Actually I meant to ask a different but related question: under what conditions isundefined a valid state in the database (as apposed to the ajax payload)? The reason that this answer is important to me is that the UI models the state of the entity (like a mini-database on the frontend), the view of the form and the ajax payload are both derived from that. So knowing the possible states of an entity helps me figure out how to create the view. (I probably didn't explain this well...)

> I seem to recall if biotype is never defined, it can stay undefined (in the database). Is that always true? Or do I need to consider the concept of cloned vs uncloned (where I have imagine the sequence name has a role)?`

There is never an `undefined` stored in the database. 
When an entity doesn't have a value for an attribute, the attribute doesn't exist.
e.g an un-cloned gene `WB1`, the database will just have: `:gene/id`, `:gene/cgc-name` and `:gene/species`. (no `:gene/sequence-name` or `:gene/biotype` attributes associated with the entity).

> If the previous value is undefined, and a undefined/null/empty value is submitted in the payload. Does that attribute gets discarded?

For the only case I'm aware of that we should permit the setting of an attribute to nothing/null/empty 
 is `cgc-name`. In the event that blanking the cgc-name is the oinly change, this will result in `:db/retract` transaction call to remove the current cgc-name.

**Comment by sibyl229** [comment]
@mgrbyte Thanks again! I know it's a lot of details here. For a seemingly minor bug, the implication of UI implementation is quite significant. So I hope to iron out the details before making changes.

I understand that "undefined" is simply a lack of tuple (or to be more specific tuple with "retracted" flag and not showing up in the current database). I just have trouble expressing questions accurately 🤔 

About cloned vs uncloned, is it possible for a gene to have **only one** of `:gene/sequence-name` and `:gene/biotype`? If so, is considered uncloned?

Does the backend to remove `null` in the payload, *when the attribute doesn't exist previously*? I'm asking because it's harder for UI code to track the previous value (as it doesn't have a DB that tracks history). I might still be able to do it, it's just a factor of consideration, if it's already implemented in the backend. 

If this needs to be implement in the UI, I'm thinking of *submitting only attributes that the user has changed* (compared to the json from their last GET). This has additional benefits outside of handling "null" . Are you okay with that? The potential concern is that is two users has edited different attributes, CAS will not raise error. This can be both an advantage and a problem depending on what is the desirable behaviour.

**Comment by mgrbyte** [comment]
_cloned_ gene: has biotype *and* sequence-name, _always_, optionally, a cgc-name.
_un-cloned_ gene: has cgc-name, never has a biotype *or* *sequence-name.

> Does the back-end to remove null in the payload, when the attribute doesn't exist previously? I'm asking because it's harder for UI code to track the previous value (as it doesn't have a DB that tracks history). I might still be able to do it, it's just a factor of consideration, if it's already implemented in the backend.

My plan was for the back-end to raise a validation error for create *and* update if the set of attributes and values do not satisfy one of either an _un-cloned_ or _cloned_ representation. 
e.g submitting a gene with a biotype _and_ an empty sequence name would fail with a 4xx response.

At present, there's a bug where the wrong spec is being used for update, which is not doing the right thing here.

In the case of _editing_ a cgc-name for gene (and only this case so far), if the cgc-name is null, then it will be "retracted". 

So in summary, genes may exist in the db (and thus be returned to the UI) in one of the following three forms:
  - cloned (always has sequence-name, biotype)
  - un-cloned (never has sequence-name, biotype, should always have cgc-name)
 - anonymous (only WBGene ID, species and status, no other attributes set)

All these variants :point_up:  have to have a id, species and status.

**Comment by sibyl229** [comment]
Thanks a lot Matt! I have a PR that gets rid of the empty biotype from the options in the gene form for clone genes. I wonder if the form for split a gene or merge gene require any change. Or shall I create a separate issue?

**Comment by mgrbyte** [comment]
> Thanks a lot Matt! I have a PR that gets rid of the empty biotype from the options in the gene form for clone genes. I wonder if the form for split a gene or merge gene require any change. Or shall I create a separate issue?

Anywhere there is a select box for biotype will need the same change I think.
Don't mind if you want a separate issue or want to re-use this one.

**Comment by sibyl229** [comment]
Whether the biotype field is required depends on if the gene is cloned. Do we allow the split-into gene to be uncloned?  The PR currently requires biotype to be set for split-from, split-into, and merge-into gene. Please let me know if this needs to be changed.

**Comment by mgrbyte** [comment]
>  Do we allow the split-into gene to be uncloned? 

No.  The `gene split` api requires a product biotype and sequence name, which means the new gene to be created can only be of the _cloned_ flavour.

> The PR currently requires biotype to be set for split-from, split-into, and merge-into gene. Please let me know if this needs to be changed

This is fine. all split and merge targets (end products) will be of the cloned flavour.
